text/0026-finality-signature-gossiping.org

```diff
@@ -0,0 +1,195 @@
+
+The central place where program logic will change is in
+[[https://github.com/xcthulhu/casper-node/blob/0a7f9e5fd7608e2f6574c1e213bd9f5e35880af5/node/src/components/linear_chain_sync.rs#L365-L368][linear_chain_sync::LinearChainSync::fetch_next_block]].
+
+#+BEGIN_SRC rust
+  State::SyncingDescendants { .. } => {
+      let next_height = block_header.height() + 1;
+      fetch_block_at_height(effect_builder, peer, next_height)
+  }
+#+END_SRC
+
+Here =fetch_block_at_height= will change to [[https://github.com/xcthulhu/casper-node/blob/0a7f9e5fd7608e2f6574c1e213bd9f5e35880af5/node/src/components/linear_chain_sync.rs#L560][fetch_block_by_hash]].
+
+Before the call to =fetch_block_by_hash=, the =LinearChainSync= process
+will need to have acquired a enough weight in signatures as discussed in
+ [[Guide-level explanation][Guide-level explanation]].
+
+A synchronizing node would make a request to its peers with a /new/
+function =fetch_finality_signatures_at_height=.
+
+When calling a peer to gossip a finality signature of height =n=, that
+peer might be faulty.  It could give a finality signature for the
+wrong height. To avoid this, the [[https://github.com/xcthulhu/casper-node/blob/0a7f9e5fd7608e2f6574c1e213bd9f5e35880af5/node/src/components/linear_chain.rs#L31-L37][FinalitySignature]] data structure
+will now include the =block_height= like so:
+
+#+begin_src rust
+/// Finality signature that can be gossiped between nodes or sent to clients.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct FinalitySignature {
+    block_hash: BlockHash,
+    block_height: u64,
+    signature: Signature,
+    public_key: PublicKey,
+}
+#+end_src
+
+The signature would be of the bytes =(block_height, block_hash)=.
+
+A faulty validator must be dealt with when syncing.  A validator could
+make finality signatures with the same height and different block
+hashes.  An approach to deal with this is as follows. First, introduce an enum:
+
+#+begin_src rust
+enum FaultyOrFetchedFinalitySignature {
+  /// A pair of finality signatures where:
+  /// - `bad_sig1.public_key == bad_sig2.public_key`
+  /// - `bad_sig1.block_height == bad_sig2.block_height`
+  /// - `bad_sig1.block_hash != bad_sig2.block_hash`
+  Faulty {
+    bad_sig1: FinalitySignature,
+    bad_sig2: FinalitySignature
+  },
+  Fetched(FinalitySignature)
+}
+#+end_src
+
+Then extend [[https://github.com/xcthulhu/casper-node/blob/0a7f9e5fd7608e2f6574c1e213bd9f5e35880af5/node/src/components/linear_chain_sync.rs#L85-L93][linear_chain_sync::State::SyncingDescendants]] with a field
+=validator_finality_signatures= with type =BTreeMap<PublicKey,
+FaultyOrFetched>=. Fetching validator signatures will update this map.
+New signatures are either added or change old signatures to faulty if
+detected.
+
+To determine the weight of a block hash in this structure: add the
+weight of non-faulty signatures with that hash /plus/ the weight of the
+faulty validators. If this is greater than 1/3 of the validator
```

Comment:
I think there are two natural ways to count:
1. The one described here: Wait until F + 1 validators have signed or been proven faulty.
2. The alternative: Wait for N - F signatures by non-faulty (as far as we know) validators.

If the assumption — at most F faulty validators — is correct, both work, and 1 is slightly faster.

If there are ≥ N - F faulty validators, both fail, of course.

But if there are strictly between F and N - F faulty validators, the difference is that 1 risks a _safety_ failure — two observers may consider conflicting blocks as finalized —, whereas 2 only risks a _liveness_ failure: You'll never collect your N - F signatures, and stall.

I know I'm the one who brought up 1; but now I wonder whether we should err on the side of safety?
Anyway, I think both options are okay; whatever you all prefer.

Reply:
> If there are ≥ N - F faulty validators, both fail, of course.

Easy enough to have either pattern and switch to the other.

Reply:
@afck after talking this over we decided we will try to detect if more than `t%` are Byzantine and stop the process if there is an equivocation catastrophe.  Did this here: https://github.com/CasperLabs/ceps/pull/26/commits/927177407952d5d24b2a245fc9030ead7eb71295